PyInstaller/building/osx.py

```diff
@@ -95,6 +95,54 @@ def _check_guts(self, data, last_build):
         # directory anyway to make sure there is no existing cruft accumulating
         return 1
 
+    def _parse_info_plist_recursive(self, plist_obj, indent=0):
+        """Recursive handling of info_plist dictionary from spec
+        Allows for handling of specification of Info.plist items that are not
+        just simple strings, for example CFBundleDocumentTypes or
+        UTExportedTypeDeclarations that contain arrays and dicts.
+        Allows handling of such hierarchical entries completely within the spec
+        file by standard python objects.
+        Accepts and converts:
+            python list to XML <array>
+            python dict to XML <dict>
+            else converts to XML <string>
+
+        Args:
+            plist_obj (dict, list, string, mix): object to parse into xml
+            indent (int): leave at default (0) for calling code, recursion uses
+                this to increase indent for each additional level of hierarchy
+
+        Returns (unicode str): info_plist_txt, xml of this item and all nested
+            items inside this item
+        """
+        info_plist_txt = ""
+
+        # make sure we aren't somehow recursing to infinity (could be reference
+        #   loop)
+        assert indent < 16
+
+        if isinstance(plist_obj, list):
+            info_plist_txt += u"\t"*indent + u"<array>\n"
+            for plist_obj_item in plist_obj:
+                info_plist_txt += self._parse_info_plist_recursive(
+                        plist_obj=plist_obj_item,
+                        indent=indent+1
+                        )
+            info_plist_txt += u"\t"*indent + u"</array>\n"
+        elif isinstance(plist_obj, dict):
+            info_plist_txt += u"\t"*indent + u"<dict>\n"
+            for k, v in plist_obj.items():
+                info_plist_txt += u"\t"*indent + u"\t<key>%s</key>\n" % k
+                info_plist_txt += self._parse_info_plist_recursive(
+                        plist_obj=v,
+                        indent=indent+1
+                        )
+            info_plist_txt += u"\t"*indent + u"</dict>\n"
+        else:
+            info_plist_txt += u"\t"*indent + u"<string>%s</string>\n" % plist_obj
```

Comment:
I can definitely do that, but my thinking here is that this falls back on the past behavior, which was just to declare anything this code can't handle as a string.

Also, is there a good Exception I could raise?  It might be nicer than bare assertions.  I wasn't sure about raising an Exception because I didn't see any in this file.

Reply:
That's a good point. Given this is current behavior, perhaps something like

``` Python
from Pyinstaller.compat import string_types
from Pyinstaller import log as logging

logger = logging.getLogger(__name__)

    # In your code
    if not isinstance(plist_obj, string_types):
        logger.warning('Provided Info.plist object %s is not a string.', plist_obj)
```

Reply:
Similarly, should I change the `assert indent < 16` to something that writes a logger.error, and then a blank `raise`, as I see for example in building/utils.py line 285?

Reply:
Also, I added your logger.warning, thanks.